src/shims/sync.rs

```diff
@@ -454,135 +456,325 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
             throw_ub_format!("destroyed a locked rwlock");
         }
 
-        rwlock_set_readers(this, rwlock_op, ScalarMaybeUndef::Undef)?;
-        rwlock_set_writers(this, rwlock_op, ScalarMaybeUndef::Undef)?;
-        rwlock_set_reader_blockset(this, rwlock_op, ScalarMaybeUndef::Undef)?;
-        rwlock_set_writer_blockset(this, rwlock_op, ScalarMaybeUndef::Undef)?;
+        rwlock_set_id(this, rwlock_op, ScalarMaybeUndef::Undef)?;
+
+        Ok(0)
+    }
+
+    fn pthread_condattr_init(&mut self, attr_op: OpTy<'tcx, Tag>) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        let default_clock_id = this.eval_libc("CLOCK_REALTIME")?;
+        condattr_set_clock_id(this, attr_op, default_clock_id)?;
+
+        Ok(0)
+    }
+
+    fn pthread_condattr_setclock(
+        &mut self,
+        attr_op: OpTy<'tcx, Tag>,
+        clock_id_op: OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        let clock_id = this.read_scalar(clock_id_op)?.not_undef()?;
+        if clock_id == this.eval_libc("CLOCK_REALTIME")?
+            || clock_id == this.eval_libc("CLOCK_MONOTONIC")?
+        {
+            condattr_set_clock_id(this, attr_op, clock_id)?;
+        } else {
+            let einval = this.eval_libc_i32("EINVAL")?;
+            return Ok(einval);
+        }
+
+        Ok(0)
+    }
+
+    fn pthread_condattr_getclock(
+        &mut self,
+        attr_op: OpTy<'tcx, Tag>,
+        clk_id_op: OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        let clock_id = condattr_get_clock_id(this, attr_op)?;
+        this.write_scalar(clock_id, this.deref_operand(clk_id_op)?.into())?;
+
+        Ok(0)
+    }
+
+    fn pthread_condattr_destroy(&mut self, attr_op: OpTy<'tcx, Tag>) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        condattr_set_clock_id(this, attr_op, ScalarMaybeUndef::Undef)?;
+
+        Ok(0)
+    }
+
+    fn pthread_cond_init(
+        &mut self,
+        cond_op: OpTy<'tcx, Tag>,
+        attr_op: OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        let attr = this.read_scalar(attr_op)?.not_undef()?;
+        let clock_id = if this.is_null(attr)? {
+            this.eval_libc("CLOCK_REALTIME")?
+        } else {
+            condattr_get_clock_id(this, attr_op)?.not_undef()?
+        };
+
+        let _ = cond_get_or_create_id(this, cond_op)?;
+        cond_set_clock_id(this, cond_op, clock_id)?;
+
+        Ok(0)
+    }
+
+    fn pthread_cond_signal(&mut self, cond_op: OpTy<'tcx, Tag>) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+        let id = cond_get_or_create_id(this, cond_op)?;
+        if let Some((thread, mutex)) = this.condvar_signal(id) {
+            reacquire_cond_mutex(this, thread, mutex)?;
+            this.unregister_timeout_callback_if_exists(thread)?;
```

Comment:
The thread might still have to wait for the mutex... are you sure it is correct to remove the timeout already?

Reply:
The documentation says that the timeout occurs if the conditional variable is not signalled in time, but I do not see a requirement for in-time acquisition of a lock:

> The pthread_cond_timedwait() function shall be equivalent to pthread_cond_wait(), except that an error is returned if the absolute time specified by abstime passes (that is, system time equals or exceeds abstime) before the condition cond is **signaled or broadcasted**, or if the absolute time specified by abstime has already been passed at the time of the call. When such timeouts occur, pthread_cond_timedwait() shall nonetheless release and re-acquire the mutex referenced by mutex, and may consume a condition signal directed concurrently at the condition variable.

Am I missing something?

Reply:
This makes sense. Please add a comment in the code explaining this.